### DISCUSSION TITLE
The public API

### DISCUSSION THREAD

**Discussion by kpagacz** [Polkas/shinyreporter#4] [Ideas]
I thought that beside the usual function call API, it would be great to have a `decorator`-based API.

I had a look at what the usual pyshiny app looks like:
```python
from shiny import ui, render, App

app_ui = ui.page_fluid(
    ui.input_slider("n", "N", 0, 100, 40),
    ui.output_text_verbatim("txt"),
)

def server(input, output, session):
    @output
    @render.text
    def txt():
        return f"n*2 is {input.n() * 2}"

# This is a shiny.App object. It must be named `app`.
app = App(app_ui, server)
```

and I was thinking it would be pretty cool if we could do something like this:
```python
from shiny import ui, render, App

@reporter_ui
app_ui = ui.page_fluid(
    ui.input_slider("n", "N", 0, 100, 40),
    ui.output_text_verbatim("txt"),
)

def server(input, output, session):
    @output
    @render.text
    @report.text(...)
    def txt():
        return f"n*2 is {input.n() * 2}"

# This is a shiny.App object. It must be named `app`.
app = App(app_ui, server)
```

Where the behaviour would be:
* the UI decorator wraps the UI into a container that has the reporter UI.
* the server decorated implicitly adds the result of the function to the report

I think that would be pretty neat and would fit nicely in the rest of the `pyshiny` ecosystem. 
We can of course do the whole functional/objective API as well.

Thoughts, @Polkas ?

**Comment by Polkas** [comment]
Hey, thanks for a great proposition.

The UI part seems a great simplification for easier usage. We still should offer direct UI injection for more control.
Somebody might want to have more control over how the UI looks like.

The server part is much more complex to answer. In an already existing solution, we are building a CARD, which is a collection of blocks. These blocks in a CARD can be a title, code, or image in any amount sb wants to have. We are adding a CARD to REPORT when the ADD CARD event is triggered. In your proposition, we have to think how to orchestrate ADD CARD like event with many outputs and/or tabs.

With the current capacity, I prefer to follow the current solution/design and later create more items to work on possible new features when people are interested in our work.
